src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/SpellCheck/SpellCheckService.cs

```diff
@@ -0,0 +1,137 @@
+    private readonly IDocumentMappingService _documentMappingService = documentMappingService;
+
+    public async Task<int[]> GetSpellCheckRangeTriplesAsync(DocumentContext documentContext, CancellationToken cancellationToken)
+    {
+        using var ranges = new PooledArrayBuilder<SpellCheckRange>();
+
+        var syntaxTree = await documentContext.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+
+        AddRazorSpellCheckRanges(ref ranges.AsRef(), syntaxTree);
+
+        var csharpRanges = await _csharpSpellCheckService.GetCSharpSpellCheckRangesAsync(documentContext, cancellationToken).ConfigureAwait(false);
+
+        if (csharpRanges.Length > 0)
+        {
+            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+            AddCSharpSpellCheckRanges(ref ranges.AsRef(), csharpRanges, codeDocument);
+        }
+
+        return ConvertSpellCheckRangesToIntTriples(ranges.ToImmutable());
+    }
+
+    private static void AddRazorSpellCheckRanges(ref PooledArrayBuilder<SpellCheckRange> ranges, RazorSyntaxTree syntaxTree)
+    {
+        // We don't want to report spelling errors in script or style tags, so we avoid descending into them at all, which
+        // means we don't need complicated logic, and it performs a bit better. We assume any C# in them will still be reported
+        // by Roslyn.
+        // In an ideal world we wouldn't need this logic at all, as we would defer to the Html LSP server to provide spell checking
+        // but it doesn't currently support it. When that support is added, we can remove all of this but the RazorCommentBlockSyntax
+        // handling.
+        foreach (var node in syntaxTree.Root.DescendantNodes(static n => n is not MarkupElementSyntax { StartTag.Name.Content: "script" or "style" }))
+        {
+            if (node is RazorCommentBlockSyntax commentBlockSyntax)
+            {
+                ranges.Add(new((int)VSInternalSpellCheckableRangeKind.Comment, commentBlockSyntax.Comment.SpanStart, commentBlockSyntax.Comment.Span.Length));
+            }
+            else if (node is MarkupTextLiteralSyntax textLiteralSyntax)
+            {
+                // Attribute names are text literals, but we don't want to spell check them because either C# will,
+                // whether they're component attributes based on property names, or they come from tag helper attribute
+                // parameters as strings, or they're Html attributes which are not necessarily expected to be real words.
+                if (node.Parent is MarkupTagHelperAttributeSyntax or
+                    MarkupAttributeBlockSyntax or
+                    MarkupMinimizedAttributeBlockSyntax or
+                    MarkupTagHelperDirectiveAttributeSyntax or
+                    MarkupMinimizedTagHelperAttributeSyntax or
+                    MarkupMinimizedTagHelperDirectiveAttributeSyntax or
+                    MarkupMiscAttributeContentSyntax)
```

Comment:
Is it worth pulling this into a helper method?

Reply:
I'd rather the compiler team pulled the tree into something that took advantage of OO, but I guess a helper method is the next best thing :D